tkg/client/upgrade_cluster.go

```diff
@@ -1115,9 +1116,13 @@ func (c *TkgClient) PatchKubernetesVersionToKubeadmControlPlane(regionalClusterC
 	}
 
 	pollOptions := &clusterclient.PollOptions{Interval: upgradePatchInterval, Timeout: upgradePatchTimeout}
-	err = regionalClusterClient.UpdateResourceWithPolling(currentKCP, clusterUpgradeConfig.KCPObjectName, clusterUpgradeConfig.KCPObjectNamespace, pollOptions)
+	kcpBytes, err := json.Marshal(currentKCP)
 	if err != nil {
-		return errors.Wrap(err, "unable to update the kubernetes version for kubeadm control plane nodes")
+		return err
+	}
+	err = regionalClusterClient.PatchResource(&capikubeadmv1beta1.KubeadmControlPlane{}, clusterUpgradeConfig.KCPObjectName, clusterUpgradeConfig.KCPObjectNamespace, string(kcpBytes), types.MergePatchType, pollOptions)
```

Comment:
Should we add a comment why we specifically patch this resource now? (vs. update before)

Q: I assume this was specifically tested and it doesn't drop a pre-existing rolloutBefore field?

Reply:
Update: looks like it works:

Hm looks like my intuition is wrong.

1 Deployed a KCP with rolloutBefore

2 `kubectl patch kcp capi-quickstart-control-plane --type merge --patch-file test.yaml --v=8`
with
```yaml
apiVersion: controlplane.cluster.x-k8s.io/v1beta1
kind: KubeadmControlPlane
metadata:
  name: capi-quickstart-control-plane
  namespace: default
spec:
  machineTemplate:
    infrastructureRef:
      apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
      kind: DockerMachineTemplate
      name: capi-quickstart-control-plane
      namespace: default
  replicas: 1
  version: v1.25.0
```
log from kubectl:
```
 Request Body: {"apiVersion":"controlplane.cluster.x-k8s.io/v1beta1","kind":"KubeadmControlPlane","metadata":{"name":"capi-quickstart-control-plane","namespace":"default"},"spec":{"machineTemplate":{"infrastructureRef":{"apiVersion":"infrastructure.cluster.x-k8s.io/v1beta1","kind":"DockerMachineTemplate","name":"capi-quickstart-control-plane","namespace":"default"}},"replicas":1,"version":"v1.25.0"}}
 ```
 
3 rolloutBefore is preserved

Reply:
It would be probably good to cover this with an envtest-based unit/integration test though

Reply:
Thanks for the check. +1 on having an envtest test.

Reply:
> Should we add a comment why we specifically patch this resource now? (vs. update before)
> 
> Q: I assume this was specifically tested and it doesn't drop a pre-existing rolloutBefore field?
> 
> (I would have guessed that a json merge patch with spec.{Version,MachineTemplate,...} would drop rolloutBefore, but I could be wrong)

Yes. rolloutBefore is immutable in CAPI. It failed to upgrade if using UpdateResource (removing rolloutBefore). So we used PatchResource instead.

Reply:
@randomvariable Wondering if that gives us enough additional signal that we don't need an additional test?
(situation will be resolved as soon as we bump to Cluster API v1.3 anyway and until then the field is immutable anyway which would lead to broken pipelines if the CLI tries to modify the field)

Reply:
Yeah, I think that's fine.
Maybe add a `TODO: @weiwenli97 Would be ideal to add an envtest unit test here` 
as it's a larger refactor to move from fakeclient to envtest.